src/peripheral/mod.rs

```diff
@@ -100,13 +104,65 @@ pub struct Cpuid {
     pub isar: [RO<u32>; 5],
     reserved1: u32,
     /// Cache Level ID
+    #[cfg(armv7m)]
     pub clidr: RO<u32>,
     /// Cache Type
+    #[cfg(armv7m)]
     pub ctr: RO<u32>,
     /// Cache Size ID
+    #[cfg(armv7m)]
     pub ccsidr: RO<u32>,
     /// Cache Size Selection
-    pub csselr: RO<u32>,
+    #[cfg(armv7m)]
+    pub csselr: RW<u32>,
+}
+
+#[cfg(armv7m)]
+const CSSELR_IND_POS: u32 = 0;
+#[cfg(armv7m)]
+const CSSELR_IND_MASK: u32 = 1 << CSSELR_IND_POS;
+#[cfg(armv7m)]
+const CSSELR_LEVEL_POS: u32 = 1;
+#[cfg(armv7m)]
+const CSSELR_LEVEL_MASK: u32 = 0x7 << CSSELR_LEVEL_POS;
+#[cfg(armv7m)]
+const CCSIDR_NUMSETS_POS: u32 = 13;
+#[cfg(armv7m)]
+const CCSIDR_NUMSETS_MASK: u32 = 0x7FFF << CCSIDR_NUMSETS_POS;
+#[cfg(armv7m)]
+const CCSIDR_ASSOCIATIVITY_POS: u32 = 3;
+#[cfg(armv7m)]
+const CCSIDR_ASSOCIATIVITY_MASK: u32 = 0x3FF << CCSIDR_ASSOCIATIVITY_POS;
+
+
+impl Cpuid {
+    /// Selects the current CCSIDR
+    ///
+    /// * `level`: the required cache level minus 1, e.g. 0 for L1, 1 for L2
+    /// * `ind`: select instruction cache (1) or data/unified cache (0)
+    #[cfg(armv7m)]
+    pub fn select_cache(&self, level: u32, ind: u32) {
+        unsafe { self.csselr.write(
+            ((level << CSSELR_LEVEL_POS) & CSSELR_LEVEL_MASK) |
+            ((ind   <<   CSSELR_IND_POS) &   CSSELR_IND_MASK)
+        )}
+    }
+
+    /// Returns the number of sets in the selected cache
+    #[cfg(armv7m)]
+    pub fn cache_num_sets(&self, level: u32, ind: u32) -> u32 {
```

Comment:
I've changed this to u16 but basically it's just meant doing a bunch of casting from u32 to u16 when we read it, and from u16 back to u32 when we use it (giving a set/way to the cache maintenance operations). Not sure it's really worthwhile?